File: STACMR-matlab/java/jMRfits.m
function [p, datafit, fits, maxbad, times] = jMRfits(nsample, y, E, shrink, reverse, proc, mrTol, seed, showStatus)
% function [p, datafit, fits, maxbad, times] = jMRfits(nsample, y, E, shrink, proc)
% jMRfits that calls the same java code as jCMRfits but with a parameter to
% use the non-coupled code (model is ignored and never set).

import au.edu.adelaide.fxmr.model.*;

if nargin<3, E={}; end
if nargin<4, shrink = -1; end
if nargin<5, reverse=false; end
if nargin<6, proc=0; end
if nargin<7, mrTol=0; end
if nargin<8, seed=-1; end
if nargin<9, showStatus=false; end

if iscell(y)
    if isstruct(y{1})
        problem = CMRxFitsProblemMaker();
        nvar = numel(y);
        ncond = numel(y{1});
    else
        problem = CMRxFitsGMProblemMaker();
        problem.setShrink(shrink);
        ngroup = size(y,1); nvar = size(y,2);
        ncond = size(y{1,1},2)*ngroup;
    end
else
    problem = CMRxFitsGMProblemMaker();
    problem.setShrink(shrink);
    ngroup = numel(unique(y(:,2)));
    nvar = numel(unique(y(:,3)));
    ncond = ngroup * (size(y,2) - 3);
end

% convert E to simpleConstraint objects
if iscell(E)
    if ~isempty(E)
        if iscell(E{1})
            %cells of cells
            for i=1:nvar
                Ecur = E{i};
                index = problem.initAdj();
                for j=1:numel(Ecur)
                    problem.addAdj(ncond, index, Ecur{j});
                end
            end
        else
            %Assume all adjs are same, from a cell
            index = problem.initAdj();
            for j=1:numel(E)
                problem.addAdj(ncond, index, E{j});
            end
            problem.dupeAdj(nvar);
        end
    end
else
    if ndims(E)==1
        %2D matrix (repeat nvar times)
        problem.addAdj(E);
        problem.dupeAdj(nvar);
    else
        %3D Matrix
        for i=1:size(E,3)
            problem.addAdj(E(:,:,i));
        end
    end
end


if iscell(y)
    if isstruct(y{1})
        ns = zeros(1,nvar);
        for ivar = 1:nvar
            problem.addMeanArray(y{ivar}.means);
            problem.addCov(y{ivar}.regcov);
            w = y{ivar}.weights;
            if isempty(w)
                w = eye(numel(y{ivar}.means));
            end
            problem.addWeightArray(w);
            if size(y{ivar}.n)==1
                ns(ivar) = y{ivar}.n;
            else
                ns(ivar) = y{ivar}.n(1,1);
            end
        end
        problem.setN(ns);
        sol = CMRxFits(nsample,problem.getProblem(),shrink,proc,false,true,mrTol,mrTol*1000,false,reverse, seed, showStatus);
    else
        for group=1:ngroup
            for var=1:nvar
                problem.addCell(group,var,y{group,var})
            end
        end
        sol = problem.solve(nsample,proc,false,true,mrTol,mrTol*1000,false,reverse,seed,showStatus);
    end
else
    problem.setGM(y);
    sol = problem.solve(nsample,proc,false,true,mrTol,mrTol*1000,false,reverse,seed,showStatus);
end

p = sol.getP();
fits = sol.getFits();
datafit = sol.getDataFit();
maxbad = zeros(1,nsample);
bads = sol.getBadnesses();
for i=1:bads.length
    maxbad(i) = bads(i).ordinal();
end
times=sol.getTimes();

File: STACMR-R/java/jMRfits.R
jMRfits <- function(nsample, y, E=list(), shrink=-1, proc=-1) {
  if (is.data.frame(y)){
    problemMaker <- new(J("au.edu.adelaide.fxmr.model.CMRxFitsGMProblemMaker"))
    problemMaker$setShrink(shrink)
    nVar <- length(unique(y[,2]))
    nCond <- length(unique(y[,3])) * (ncol(y) - 3)
  }else{    
    if (!is.null(y[[1]][["means"]])){
      #Parametric case, probably passed in from staSTATS
      problemMaker <- new(J("au.edu.adelaide.fxmr.model.CMRxFitsProblemMaker"))
      nVar <- length(y)
      nCond <- length(y[[1]]$means)
    }else{
      problemMaker <- new(J("au.edu.adelaide.fxmr.model.CMRxFitsGMProblemMaker"))
      problemMaker$setShrink(shrink)
      nVar <- length(y)
      nCond <- length(y[[1]])
    }
  }

  if (!missing("E") && length(E) > 0) {
    if (is.list(E) && is.list(E[[1]])){
      #3d list, different constrains for each variable
      nE <- length(E)
      for(iVar in 1:nVar){
        Ecur <- E[[iVar]]
        index <- problemMaker$initAdj()
        if (length(Ecur) > 0){
          for (j in 1:length(Ecur)){
            problemMaker$addAdj(as.integer(nCond), index, as.integer(Ecur[[j]]))
          }
        }
      }
    }else if (is.list(E)){
      #Same list for all variables
      index = problemMaker$initAdj();
      for (j in 1:length(E)){
        problemMaker$addAdj(as.integer(nCond), index, as.integer(E[[j]]));
      }
      problemMaker$dupeAdj(nVar);
    }
  }
  
  if (is.data.frame(y)){
    problemMaker$setGM(.jarray(as.matrix(y), dispatch=T));
    sol <- problemMaker$solve(as.integer(nsample),as.integer(proc),FALSE,TRUE);
  }else{
    if (!is.null(y[[1]][["means"]])){
      #Parametric case, probably passed in from staSTATS
      ns = matrix(1,nVar)
      for (iVar in 1:nVar){
        problemMaker$addMeanArray(y[[iVar]]$means)
        w <- y[[iVar]]$weights
        if (is.null(w)){
          w <- diag(length(y[[iVar]]$means));
        }
        problemMaker$addWeightArray(.jarray(w, dispatch = T))
        problemMaker$addCov(.jarray(y[[iVar]]$regcov, dispatch = T))
        if (length(y[[iVar]]$n) == 1){
          ns[[iVar]] <- y[[iVar]]$n
        }else{
          ns[[iVar]] <- y[[iVar]]$n[[1]][[1]]
        }
      }
      problemMaker$setN(as.integer(ns))
      problem <- problemMaker$getProblem()
      
      sol <- new(J("au.edu.adelaide.fxmr.model.CMRxFits"),as.integer(nsample),problem,shrink,as.integer(proc))
    }else{
      for (group in 1:nCond){
        for (iVar in 1:nVar){
          problemMaker$addCell(as.integer(iVar),as.integer(group),.jarray(as.matrix(y[[iVar]][[group]]), dispatch = T))
        }
      }
      sol <- problemMaker$solve(as.integer(nsample),as.integer(proc),FALSE,TRUE);
    }
  }
  
  return(list(p = sol$getP(), fits = sol$getFits(),datafit = sol$getDataFit()))
}